src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,94 @@
+from collections import deque
 from typing import Optional, Type
 
-from ape.exceptions import ContractLogicError, TransactionError
+import ape
+from ape.exceptions import APINotImplementedError, ContractLogicError, TransactionError
 
 
 class RevertsContextManager:
-    def __init__(self, expected_message: Optional[str] = None):
+    def __init__(self, expected_message: Optional[str] = None, dev_message: Optional[str] = None):
         self.expected_message = expected_message
+        self.dev_message = dev_message
 
-    def __enter__(self):
-        pass
+    def _check_dev_message(self, exception: ContractLogicError):
+        """
+        Attempts to extract a dev-message from the contract source code by inspecting what
+        instruction(s) led to a transaction revert.
 
-    def __exit__(self, exc_type: Type, exc_value: Exception, traceback) -> bool:
-        if exc_type is None:
-            raise AssertionError("Transaction did not revert.")
+        Raises:
+            AssertionError: When the trace or source can not be retrieved, the dev message cannot
+            be found, or the found dev message does not match the expected dev message.
+        """
+        receipt = exception.receipt
```

Comment:
Instead of storing the receipt on the `exception`, we could potentially look it up now and only store the `txn_hash` on the receipt.

We could make `RevertsContextManager` use the `ManagerAccessMixin` and do

```python
receipt = self.chain_manager.accounts_history[exception.txn_hash]
```

That may help with the refactor you were asking about, during the parts where we get the receipt during `get_vm_err()`

It may also help with other comments I am adding.

Thoughts?

Reply:
Updated in 5d4a7c24 but to pass `TransactionAPI` instead, because we basically can get everything we need (`receiver`, `txn_hash`, `provider`, etc) from there

So `TransactionError` now has an optional `txn` parameter for its constructor - I only supply `txn` during failed `Web3Provider.send_transaction()` calls but I'm not sure if we need it elsewhere. For now it basically only facilitates this dev-message assertion feature